src/sync/sync_id.rs

```diff
@@ -0,0 +1,58 @@
+use crate::util::uuid;
+use std::collections::HashMap;
+use std::sync::atomic::{AtomicU32, Ordering};
+use std::sync::Mutex;
+use str_macro::str;
+
+lazy_static! {
+    static ref SESSION_ID: String = {
+        let mut randoms: [u8; 4] = [0; 4];
+        uuid::make_random_numbers(&mut randoms);
+        format!(
+            "{:x}{:x}{:x}{:x}",
+            randoms[0], randoms[1], randoms[2], randoms[3]
+        )
+    };
+    static ref SYNC_COUNTERS: Mutex<HashMap<String, AtomicU32>> = Mutex::new(HashMap::new());
+}
+
+// new() returns a new syncid of the form <clientid>-<sessionid>-<sync count>.
+// The sync count enables one to find the sync following or preceeding a given
+// sync. The sessionid scopes the sync count, ensuring the syncid is
+// probabilistically unique across restarts (which is good enough).
+pub fn new(client_id: &str) -> String {
+    let mut map = match SYNC_COUNTERS.lock() {
+        Err(e) => {
+            error!("", "error getting sync count: {:?}", e);
+            return str!("busted-sync-counter");
+        }
+        Ok(m) => m,
+    };
+    let counter = match map.get(client_id) {
+        None => {
+            let counter = AtomicU32::new(0);
+            map.insert(client_id.into(), counter);
+            map.get(client_id).unwrap()
```

Comment:
Do we really need to insert and get?

Reply:
yes. it gets moved into the map in 34 so we need to get it back out in 35.